build_tools/python/e2e_test_framework/serialization.py

```diff
@@ -0,0 +1,249 @@
+
+  raise ValueError(f"Unsupported object: {obj}.")
+
+
+def _deserialize(data,
+                 obj_type: Type,
+                 keyed_obj_map: Dict[str, Any],
+                 obj_cache: Dict[str, Any] = {}):
+  """Deserializes the data back to the typed object.
+
+  Args:
+    data: serialized data.
+    obj_type: type of the data.
+    keyed_obj_map: container of the keyed serializable object.
+  Returns:
+    A deserialized object.
+  """
+
+  deserialize_func = getattr(obj_type, DESERIALIZE_FUNC_NAME, None)
+  if deserialize_func is not None:
+    return deserialize_func(data, keyed_obj_map, obj_cache)
+
+  elif typing.get_origin(obj_type) == list:
+    subtype, = typing.get_args(obj_type)
+    return [
+        _deserialize(item, subtype, keyed_obj_map, obj_cache) for item in data
+    ]
+
+  elif typing.get_origin(obj_type) == collections.OrderedDict:
+    # Only supports OrderedDict as it provides the clear deterministic semantic.
+    _, value_type = typing.get_args(obj_type)
+    return collections.OrderedDict(
+        (key, _deserialize(value, value_type, keyed_obj_map, obj_cache))
+        for key, value in data.items())
+
+  elif typing.get_origin(obj_type) == Union:
+    subtypes = typing.get_args(obj_type)
+    if len(subtypes) != 2 or NONE_TYPE not in subtypes:
+      raise ValueError(f"Unsupported union type: {obj_type}.")
+    subtype = subtypes[0] if subtypes[1] == NONE_TYPE else subtypes[1]
+    return _deserialize(data, subtype, keyed_obj_map, obj_cache)
+
+  elif issubclass(obj_type, Enum):
+    for member in obj_type:
+      if data == member.name:
+        return member
+    raise ValueError(f"Member {data} not found in the enum {obj_type}.")
+
+  return data
+
+
+def serializable(cls=None,
+                 keyed_obj: bool = False,
+                 type_key: Optional[str] = None,
+                 id_field: str = "id"):
+  """Decorator to make a dataclass serializable.
+  
+  Args:
+    keyed_obj: is the class a keyed object, which is unique per id and will only
+      have one copy in the serialization per id.
+    type_key: string defining the object type, must be set if keyed_obj is True.
+    id_field: field name of the id field of a keyed object.
+
+  Example:
+    @serializable
+    @dataclass
+    class A(object):
+      ...
+
+    @serialzable(keyed_obj=True, type_key="obj_b")
+    @dataclass
+    class B(object):
+      id: str
+  """
+
+  if keyed_obj and type_key is None:
+    raise ValueError("type_key must be set if keyed_by_id is true.")
+  if type_key is not None and ":" in type_key:
+    raise ValueError("':' is the reserved character in type_key.")
+
+  def wrap(cls):
+    if not dataclasses.is_dataclass(cls):
+      raise ValueError(f"{cls} is not a dataclass.")
+
+    fields = dataclasses.fields(cls)
+
+    def serialize(self, keyed_obj_map: OrderedDict[str, Any]):
+      if not keyed_obj:
+        return _fields_to_dict(self, fields, keyed_obj_map)
+
+      obj_id = getattr(self, id_field)
```

Comment:
Is it possible to validate in the main body of the function that the class has such a field. Also could check that it's a dataclass. I think this doesn't need to be deferred.

Reply:
Done.